src/node_sqlite.cc

```diff
@@ -128,6 +149,169 @@ inline void THROW_ERR_SQLITE_ERROR(Isolate* isolate, int errcode) {
+                     std::string source_db,
+                     std::string destination_name,
+                     std::string dest_db,
+                     int pages,
+                     Local<Function> progressFunc)
+      : ThreadPoolWork(env, "node_sqlite3.BackupJob"),
+        env_(env),
+        source_(source),
+        source_db_(source_db),
+        destination_name_(destination_name),
+        dest_db_(dest_db),
+        pages_(pages) {
+    resolver_.Reset(env->isolate(), resolver);
+    progressFunc_.Reset(env->isolate(), progressFunc);
+  }
+
+  void ScheduleBackup() {
+    Isolate* isolate = env()->isolate();
+    HandleScope handle_scope(isolate);
+
+    backup_status_ = sqlite3_open(destination_name_.c_str(), &pDest_);
+
+    Local<Promise::Resolver> resolver =
+        Local<Promise::Resolver>::New(env()->isolate(), resolver_);
+
+    Local<Object> e = Local<Object>();
+
+    if (backup_status_ != SQLITE_OK) {
+      e = CreateSQLiteError(isolate, pDest_).ToLocalChecked();
+
+      Cleanup();
+
+      resolver->Reject(env()->context(), e).ToChecked();
+
+      return;
+    }
+
+    pBackup_ = sqlite3_backup_init(
+        pDest_, dest_db_.c_str(), source_->Connection(), source_db_.c_str());
+
+    if (pBackup_ == nullptr) {
+      e = CreateSQLiteError(isolate, pDest_).ToLocalChecked();
+
+      sqlite3_close(pDest_);
+
+      resolver->Reject(env()->context(), e).ToChecked();
+
+      return;
+    }
+
+    this->ScheduleWork();
+  }
+
+  void DoThreadPoolWork() override {
+    backup_status_ = sqlite3_backup_step(pBackup_, pages_);
+  }
+
+  void AfterThreadPoolWork(int status) override {
+    HandleScope handle_scope(env()->isolate());
+
+    if (resolver_.IsEmpty()) {
+      Cleanup();
+
+      return;
+    }
+
+    Local<Promise::Resolver> resolver =
+        Local<Promise::Resolver>::New(env()->isolate(), resolver_);
+
+    if (!(backup_status_ == SQLITE_OK || backup_status_ == SQLITE_DONE ||
+          backup_status_ == SQLITE_BUSY || backup_status_ == SQLITE_LOCKED)) {
+      Local<Object> e =
+          CreateSQLiteError(env()->isolate(), backup_status_).ToLocalChecked();
+
+      Cleanup();
+
+      resolver->Reject(env()->context(), e).ToChecked();
+
+      return;
+    }
+
+    int total_pages = sqlite3_backup_pagecount(pBackup_);
+    int remaining_pages = sqlite3_backup_remaining(pBackup_);
+
+    if (remaining_pages != 0) {
+      Local<Function> fn =
+          Local<Function>::New(env()->isolate(), progressFunc_);
+
+      if (!fn.IsEmpty()) {
+        Local<Value> argv[] = {
+            Integer::New(env()->isolate(), total_pages),
+            Integer::New(env()->isolate(), remaining_pages),
+        };
+
+        TryCatch try_catch(env()->isolate());
+        fn->Call(env()->context(), Null(env()->isolate()), 2, argv)
+            .FromMaybe(Local<Value>());
+
+        if (try_catch.HasCaught()) {
+          Cleanup();
+
+          resolver->Reject(env()->context(), try_catch.Exception()).ToChecked();
```

Comment:
Probably best to check `try_catch.CanContinue()` before calling `Reject` to check for fatal errors that are terminating the isolate.

Reply:
If we do that I think we need to let somehow the user know that the backup is happening even though the `progress` function is not called.

I added the `CanContinue` to check and ran the test below.

```js
    await database.backup('database.bkp', {
      progress: () => {.throw 'something went wrong'; }
    });
```

For this `progress` function the `CanContinue` is true. The backup is finished but there's no information about it.

Maybe we can emit a warning or something?